fw/tls.c

```diff
@@ -820,6 +822,22 @@ tfw_tls_sni(TlsCtx *ctx, const unsigned char *data, size_t len)
 				 PR_TFW_STR(&srv_name));
 			return -ENOENT;
 		}
+
+		/* Save server name for later validation with authority from requests */
+		if (likely(len <= sizeof(ctx->sni_buf))) {
+			ctx->sni_len = len;
+			ctx->sni = ctx->sni_buf;
+			memcpy_fast(ctx->sni_buf, data, len);
+		} else {
+			ctx->sni = kmalloc(len, GFP_KERNEL | GFP_NOWAIT);
+			if (unlikely(ctx->sni == NULL)) {
+				/* this is non-fatal, just warn about it */
+				SNI_WARN("Unable to allocate %zu bytes for server_name", len);
```

Comment:
That's not practical to go with host names larger than 4KB, so let's return an error code and close the connection